File: attentive-android-sdk/src/main/java/com/attentive/androidsdk/AttentiveSettingsService.java
package com.attentive.androidsdk;

import static com.attentive.androidsdk.internal.util.AppInfo.isDebuggable;

import android.app.Service;
import android.content.Intent;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import com.attentive.androidsdk.internal.util.LightTree;
import timber.log.Timber;

public class AttentiveSettingsService extends Service {

    public static final String EXTRA_SET_SKIP_FATIGUE = "setSkipFatigue";
    public static final String EXTRA_RESET_SETTINGS = "resetSettings";
    public static final String EXTRA_SET_LOG_LEVEL = "setLogLevel";

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (!isDebuggable(this)) {
            Timber.e("onCreate: not in debug mode");
            return Service.START_STICKY;
        }
        final Bundle extras = intent.getExtras();
        if (extras == null) {
            Timber.d("onCreate: extras is null");
            return Service.START_STICKY;
        }

        final SettingsService settingsService = ClassFactory.buildSettingsService(ClassFactory.buildPersistentStorage(getApplicationContext()));
        handleSkipFatigueExtra(extras, settingsService);
        handleResetSettingsExtra(extras, settingsService);
        handleSetLogLevelExtra(extras, settingsService);
        stopSelf();
        return Service.START_STICKY;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        if (isDebuggable(this)) {
            Timber.plant(new Timber.DebugTree());
        } else {
            Timber.plant(new LightTree());
        }
        return null;
    }

    @VisibleForTesting
    public static void handleSkipFatigueExtra(Bundle extras, SettingsService settingsService) {
        if (extras.containsKey(EXTRA_SET_SKIP_FATIGUE)) {
            Timber.d("Setting skip fatigue...");
            boolean skipFatigue = extras.getBoolean(EXTRA_SET_SKIP_FATIGUE);
            settingsService.setSkipFatigueEnabled(extras.getBoolean(EXTRA_SET_SKIP_FATIGUE, false));
            Timber.i("skipFatigue set to: %s", skipFatigue);
        }
    }

    @VisibleForTesting
    public static void handleResetSettingsExtra(Bundle extras, SettingsService settingsService) {
        if (extras.containsKey(EXTRA_RESET_SETTINGS) && extras.getBoolean(EXTRA_RESET_SETTINGS, false)) {
            Timber.d("Resetting settings...");
            settingsService.resetSettings();
            Timber.i("Settings reset");
        }
    }

    @VisibleForTesting
    public static void handleSetLogLevelExtra(Bundle extras, SettingsService settingsService) {
        if (extras.containsKey(EXTRA_SET_LOG_LEVEL)) {
            Timber.d("Setting log level...");
            int logLevel = extras.getInt(EXTRA_SET_LOG_LEVEL, -1);
            AttentiveLogLevel attentiveLogLevel = AttentiveLogLevel.fromId(logLevel);
            if (attentiveLogLevel == null) {
                Timber.w("Log level should be one of: %s", (Object[]) AttentiveLogLevel.values());
                Timber.i("Log level not set");
                return;
            }
            settingsService.setLogLevel(attentiveLogLevel);
            Timber.i("log level set to: %s", attentiveLogLevel);
        }
    }
}

File: attentive-android-sdk/src/main/java/com/attentive/androidsdk/AttentiveSettingsService.kt
package com.attentive.androidsdk

import android.app.Service
import android.content.Intent
import android.os.Bundle
import android.os.IBinder
import androidx.annotation.VisibleForTesting
import com.attentive.androidsdk.internal.util.AppInfo
import timber.log.Timber
import timber.log.Timber.DebugTree

class AttentiveSettingsService : Service() {
    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        if (!AppInfo.isDebuggable(this)) {
            Timber.e("onCreate: not in debug mode")
            return START_STICKY
        }
        val extras = intent.extras
        if (extras == null) {
            Timber.d("onCreate: extras is null")
            return START_STICKY
        }

        val settingsService = ClassFactory.buildSettingsService(
            ClassFactory.buildPersistentStorage(
                applicationContext
            )
        )
        handleSkipFatigueExtra(extras, settingsService)
        handleResetSettingsExtra(extras, settingsService)
        handleSetLogLevelExtra(extras, settingsService)
        stopSelf()
        return START_STICKY
    }

    override fun onBind(intent: Intent): IBinder? {
        Timber.plant(DebugTree())
        return null
    }

    companion object {
        const val EXTRA_SET_SKIP_FATIGUE: String = "setSkipFatigue"
        const val EXTRA_RESET_SETTINGS: String = "resetSettings"
        const val EXTRA_SET_LOG_LEVEL: String = "setLogLevel"

        @JvmStatic
        @VisibleForTesting
        fun handleSkipFatigueExtra(extras: Bundle, settingsService: SettingsService) {
            if (extras.containsKey(EXTRA_SET_SKIP_FATIGUE)) {
                Timber.d("Setting skip fatigue...")
                val skipFatigue = extras.getBoolean(EXTRA_SET_SKIP_FATIGUE, false)
                settingsService.isSkipFatigueEnabled = skipFatigue
                Timber.i("skipFatigue set to: %s", skipFatigue)
            }
        }

        @JvmStatic
        @VisibleForTesting
        fun handleResetSettingsExtra(extras: Bundle, settingsService: SettingsService) {
            if (extras.containsKey(EXTRA_RESET_SETTINGS) && extras.getBoolean(
                    EXTRA_RESET_SETTINGS,
                    false
                )
            ) {
                Timber.d("Resetting settings...")
                settingsService.resetSettings()
                Timber.i("Settings reset")
            }
        }

        @VisibleForTesting
        fun handleSetLogLevelExtra(extras: Bundle, settingsService: SettingsService) {
            if (extras.containsKey(EXTRA_SET_LOG_LEVEL)) {
                Timber.d("Setting log level...")
                val logLevel = extras.getInt(EXTRA_SET_LOG_LEVEL, -1)
                val attentiveLogLevel =
                    AttentiveLogLevel.fromId(logLevel)
                if (attentiveLogLevel == null) {
                    Timber.w(
                        "Log level should be one of: %s",
                        *AttentiveLogLevel.entries.toTypedArray() as Array<Any?>
                    )
                    Timber.i("Log level not set")
                    return
                }
                settingsService.logLevel = attentiveLogLevel
                Timber.i("log level set to: %s", attentiveLogLevel)
            }
        }
    }
}
